From: Benjamin Drung <benjamin.drung@canonical.com>
Date: Thu, 3 Sep 2026 11:41:18 +0200
Subject: feat(dracut): add --prefer option

Add a `--prefer` option and a `prefer_dracutmodules` config option to
specify Dracut modules that should be included if their preconditions
are met.

Origin: upstream, https://github.com/dracut-ng/dracut/commit/4b8c11c0590b546e37908968042eaf7c7d669b17
---
 dracut.sh                    | 18 ++++++++++++++++--
 man/dracut.8.adoc            | 16 ++++++++++++++++
 man/dracut.conf.5.adoc       |  4 ++++
 shell-completion/bash/dracut |  4 ++--
 4 files changed, 38 insertions(+), 4 deletions(-)

diff --git a/dracut.sh b/dracut.sh
index 5f149e0..fa076aa 100755
--- a/dracut.sh
+++ b/dracut.sh
@@ -96,6 +96,8 @@ Creates initial ramdisk images for preloading modules
   --force-add [LIST]    Force to add a space-separated list of dracut modules
                          to the default set of modules, when -H is specified.
   -o, --omit [LIST]     Omit a space-separated list of dracut modules.
+  --prefer [LIST]       Specify a space-separated list of dracut modules that
+                         will be added if their preconditions are met.
   -m, --modules [LIST]  Specify a space-separated list of dracut modules to
                          call when building the initramfs. Modules are located
                          in /usr/lib/dracut/modules.d.
@@ -392,6 +394,7 @@ rearrange_params() {
             --long add-drivers: \
             --long force-drivers: \
             --long omit-drivers: \
+            --long prefer: \
             --long modules: \
             --long omit: \
             --long drivers: \
@@ -595,6 +598,7 @@ no_kernel=
 nofscks=
 omit_dracutmodules=
 omit_drivers=
+prefer_dracutmodules=
 prefix=
 ro_mnt=
 squash_compress=
@@ -696,6 +700,11 @@ while :; do
                     PARMS_TO_STORE+=" '$2'"
                     shift
                     ;;
+                --prefer)
+                    prefer_dracutmodules_l+=("$2")
+                    PARMS_TO_STORE+=" '$2'"
+                    shift
+                    ;;
                 -d | --drivers)
                     drivers_l+=("$2")
                     PARMS_TO_STORE+=" '$2'"
@@ -1126,6 +1135,7 @@ export SYSTEMCTL=${SYSTEMCTL:-systemctl}
 ((${#add_dracutmodules_l[@]})) && add_dracutmodules+=" ${add_dracutmodules_l[*]} "
 ((${#omit_dracutmodules_l[@]})) && omit_dracutmodules+=" ${omit_dracutmodules_l[*]} "
 ((${#force_add_dracutmodules_l[@]})) && force_add_dracutmodules+=" ${force_add_dracutmodules_l[*]} "
+((${#prefer_dracutmodules_l[@]})) && prefer_dracutmodules+=" ${prefer_dracutmodules_l[*]} "
 ((${#fscks_l[@]})) && fscks+=" ${fscks_l[*]} "
 ((${#add_fstab_l[@]})) && add_fstab+=" ${add_fstab_l[*]} "
 ((${#install_items_l[@]})) && install_items+=" ${install_items_l[*]} "
@@ -1863,7 +1873,7 @@ check_module() {
         fi
     fi
 
-    if [[ " $dracutmodules $add_dracutmodules $force_add_dracutmodules" == *\ $_mod\ * ]]; then
+    if [[ " $dracutmodules $add_dracutmodules $force_add_dracutmodules $prefer_dracutmodules" == *\ $_mod\ * ]]; then
         if [[ " $dracutmodules $force_add_dracutmodules " == *\ $_mod\ * ]]; then
             module_check "$_mod" 1 "$_moddir"
             _ret=$?
@@ -1871,7 +1881,7 @@ check_module() {
             module_check "$_mod" 0 "$_moddir"
             _ret=$?
         fi
-        # explicit module, so also accept _ret=255
+        # explicit or preferred module, so accept _ret=0 and _ret=255
         [[ $_ret == 0 || $_ret == 255 ]] || return 1
     else
         # module not in our list
@@ -1897,6 +1907,9 @@ check_module() {
         [[ " $add_dracutmodules " == *\ $_mod\ * ]] \
             && [[ " $add_dracutmodules " != *\ $_moddep\ * ]] \
             && add_dracutmodules+=" $_moddep "
+        [[ " $prefer_dracutmodules " == *\ $_mod\ * ]] \
+            && [[ " $prefer_dracutmodules " != *\ $_moddep\ * ]] \
+            && prefer_dracutmodules+=" $_moddep "
         [[ " $force_add_dracutmodules " == *\ $_mod\ * ]] \
             && [[ " $force_add_dracutmodules " != *\ $_moddep\ * ]] \
             && force_add_dracutmodules+=" $_moddep "
@@ -2600,6 +2613,7 @@ if [[ $printconfig ]]; then
         nofscks \
         omit_dracutmodules \
         omit_drivers \
+        prefer_dracutmodules \
         prefix \
         ro_mnt \
         squash_compress \
diff --git a/man/dracut.8.adoc b/man/dracut.8.adoc
index 2f442b4..8b39aab 100644
--- a/man/dracut.8.adoc
+++ b/man/dracut.8.adoc
@@ -117,6 +117,22 @@ example:
 ----
 ===============================
 
+**--prefer** _<list of dracut modules>_::
+    Add a space-separated list of dracut modules to be preferred. Preferred modules
+    will be included if they are supported by your system, but safely skipped if
+    they are not compatible or have missing dependencies.
+    If a module appears in both _prefer_ and _omit_ lists, it will be omitted.
+    This parameter can be specified multiple times.
++
+[NOTE]
+===============================
+If the list has multiple arguments, then you have to put these in quotes. For
+example:
+----
+# dracut --prefer "module1 module2"  ...
+----
+===============================
+
 **-m, --modules** _<list of dracut modules>_::
     Specify a space-separated list of dracut modules to call when building the
     initramfs. Modules are located in _/usr/lib/dracut/modules.d_. This
diff --git a/man/dracut.conf.5.adoc b/man/dracut.conf.5.adoc
index c8e8360..32ec939 100644
--- a/man/dracut.conf.5.adoc
+++ b/man/dracut.conf.5.adoc
@@ -50,6 +50,10 @@ WARNING: Avoid manually omitting dracut modules, as you may inadvertently
 remove essential ones that dracut can't detect or warn you about. Using this
 option is not recommended and is at your own risk.
 
+*prefer_dracutmodules+=*" __<dracut modules>__ "::
+Specify a space-separated list of dracut modules that will be added
+if their preconditions are met.
+
 *dracutmodules+=*" __<dracut modules>__ "::
 Specify a space-separated list of dracut modules to call when building the
 initramfs. Modules are located in _/usr/lib/dracut/modules.d_.
diff --git a/shell-completion/bash/dracut b/shell-completion/bash/dracut
index c1388ab..ccc3d48 100644
--- a/shell-completion/bash/dracut
+++ b/shell-completion/bash/dracut
@@ -40,7 +40,7 @@ _dracut() {
             --no-uefi --no-ukify --no-machineid --version --parallel --printconfig
             '
         [ARG]='-a -m -o -d -I -k -c -L -r -i
-            --kver --add --force-add --add-drivers --force-drivers
+            --kver --add --force-add --add-drivers --force-drivers --prefer
             --omit-drivers --modules --omit --drivers --filesystems --install
             --fwdir --libdirs --fscks --add-fstab --mount --device
             --kmoddir --conf --confdir --tmpdir --stdlog --compress --prefix
@@ -63,7 +63,7 @@ _dracut() {
                 comps=$(compgen -f -- "$cur")
                 compopt -o filenames
                 ;;
-            -a | -m | -o | --add | --modules | --omit)
+            -a | -m | -o | --add | --modules | --omit | --prefer)
                 comps=$(dracut --list-modules 2> /dev/null)
                 ;;
             --compress-level)
